stellargraph/data/explorer.py

```diff
@@ -780,3 +783,166 @@ def _check_neighbourhood_sizes(self, in_size, out_size):
+                * None (default) - the initial edge is picked from a uniform distribution
+                * "exponential" - heavily biased towards more recent edges.
+            walk_bias (str, optional): distribution to use when choosing a random
+                neighbour to walk through. Available options are:
+                * None (default) - neighbours are picked from a uniform distribution
+                * "exponential" - exponentially decaying probability, resulting in a bias
+                    towards shorter time gaps
+            p_walk_success_threshold (float): Lower bound for the proportion of successful
+                (i.e. longer than minimum length) walks. If the 95% percentile of the
+                estimated proportion is less than the provided threshold, a RuntimeError
+                will be raised.
+            seed (int, optional): Random number generator seed; default is None
+
+        Returns:
+            List of lists of node ids for each of the random walks
+
+        """
+        if cw_size < 2:
+            raise ValueError(
+                f"cw_size: context window size should be greater than 1, found {cw_size}"
+            )
+        if max_walk_length < cw_size:
+            raise ValueError(
+                f"max_walk_length: maximum walk length should not be less than the context window size, found {max_walk_length}"
+            )
+
+        np_rs = self._np_random_state if seed is None else np.random.RandomState(seed)
+        walks = []
+        num_cw_curr = 0
+
+        edges, times = self.graph.edges(include_edge_weight=True)
+        edge_biases = self._temporal_biases(
+            times, None, bias_type=initial_edge_bias, is_forward=False,
+        )
+
+        successes = 0
+        failures = 0
+
+        def not_progressing_enough():
+            # Estimate the probability p of a walk being long enough; the 95% percentile is used to
+            # be more stable with respect to randomness. This uses Beta(1, 1) as the prior, since
+            # it's uniform on p
+            posterior = stats.beta.ppf(0.95, 1 + successes, 1 + failures)
+            return posterior < p_walk_success_threshold
+
+        # loop runs until we have enough context windows in total
+        while num_cw_curr < num_cw:
+            first_edge_index = self._sample(len(edges), edge_biases, np_rs)
+            src, dst = edges[first_edge_index]
+            t = times[first_edge_index]
+
+            remaining_length = num_cw - num_cw_curr + cw_size - 1
+
+            walk = self._walk(
+                src, dst, t, min(max_walk_length, remaining_length), walk_bias, np_rs
+            )
+            if len(walk) >= cw_size:
+                walks.append(walk)
+                num_cw_curr += len(walk) - cw_size + 1
+                successes += 1
+            else:
+                failures += 1
+                if not_progressing_enough():
+                    raise RuntimeError(
```

Comment:
I would suggest that this should be a warning message rather than raising an exception. We should let the user know that his choices are causing the algorithm to run for too long. Too many failures does not necessarily mean that if one waits long enough the walker will not terminate with enough data to continue training the model. We should allow users to keep going if they so wish as we cannot predict the kind of datasets and the special circumstances people are working with.

Reply:
> We should allow users to keep going if they so wish as we cannot predict the kind of datasets and the special circumstances people are working with.

This is the reasoning behind the parameter that you've asked about in your [other comment](https://github.com/stellargraph/stellargraph/pull/787/files#r376828022) (`p_walk_success_threshold`). I can make the documentation more clear on that parameter, but I don't really see the value in turning turn this error into a warning instead. The parameter exists so that if they see this error, and want to ignore it, they can reduce that value to however low they want to go (even zero I guess if they really want an infinite loop). I think it's *much* more likely that when someone hits this error, it's indicative of something actually wrong (wrong setup, wrong parameters) that should be fixed.

Reply:
The current threshold implies that this will only fail if less 1% of the attempted random walks are successful. This is effectively a 100&times; slowdown, and, as Kevin says, probably indicates some deeper problem. Prior art in this space: `sklearn` models only emit a `ConvergenceWarning` warning, not raise an error, if fitting model fails to converge after a (user-specified) number of iterations (the `max_iter` parameter), but it does not go into an infinite/very slow loop, and instead returns the partially-trained model.

I think it's quite bad to go into an infinite loop, for anything other than interactive experimentation. If a process isn't being actively monitored (e.g. batch processing on a server/cluster), a infinite loop could translate into hundreds or thousands of dollars wasted on useless compute, along with whatever the consequences are of not finishing whatever task is being attempted; at least with a hard error, it's flagged in an obvious manner.

An approach more similar to the `sklearn` approach that still avoids catastrophic failures for misconfiguration (or, at least, very poor configuration) would be to emit a warning but still return as many walks as were performed, e.g.:

```python
def run(...):
    ...
    while num_cw_curr < num_cw:
        ...
       if not_progressing_enough():
            warnings.warn(...)
            # there's been too many non-productive walks, so give up and go with what's been found up to this point
            break
    return walks
```
Or, something that might be more intuitive than estimating the probability of a walk working would be to put a hard cap on the number of walks that are performed:

```python
def run(
    num_cw,
    ...
    maximum_walks_multiplier=100 # do at most 100*num_cw random walks
):
    ...
    maximum_number_of_walks = num_cw * maximum_walks_multiplier

    while num_cw_curr < num_cw:
        ...

        if success + failures >= maximum_number_of_walks:
            warnings.warn(...)
            # there's been too many non-productive walks, so give up and go with what's been found up to this point
            break

    return walks
```

Reply:
> Prior art in this space: sklearn models only emit a ConvergenceWarning warning, not raise an error, if fitting model fails to converge after a (user-specified) number of iterations (the max_iter parameter), but it does not go into an infinite/very slow loop, and instead returns the partially-trained model.

> An approach more similar to the sklearn approach that still avoids catastrophic failures for misconfiguration (or, at least, very poor configuration) would be to emit a warning but still return as many walks as were performed,

Hmmmm, I think I'm still a bit unconvinced. I'd think it's currently a good thing that this method always returns `num_cw` number of context windows, especially given that the current use-case it's implemented for is generating context pairs as training examples. Having to catch cases where you end up with fewer training example seems like a hassle, and not really something the user would intentionally want to do..?

> Or, something that might be more intuitive than estimating the probability of a walk working would be to put a hard cap on the number of walks that are performed

Hmm yeah could be reasonable, but then we kind of lose the early stopping (e.g. if the graph is large and so `num_cw` is large, it could still run for ages without collecting a single successful walk), so I'm not sure if it's definitely an improvement 🤷‍♂